### ISSUE TITLE
Dependency graph is not correctly computed with go 1.18

### ISSUE THREAD

**Issue by mateuszlewko** [nx-go/nx-go#66]
Assume that we have a go app depending on a go library in one workspace. Running `nx graph` doesn't show dependency between them. I think it's because you're trying to read `.Module` from the output of `go list -json`. The output for that command run with go1.18 doesn't contain `Module` property.

**Comment by phillipCouto** [comment]
@beeman I am more than happy to look into this if you would like.

**Comment by beeman** [comment]
> @beeman I am more than happy to look into this if you would like.

That would be awesome @phillipCouto 🙌

**Comment by dustyhorizon** [comment]
Dependency graphs in go1.18 are not fully fixed as of nx-go 2.5.0.

go1.18 introduces workspaces which produces multiple modules when running `go list -m -json`

An example output of this is as follows:
```
$ go list -m -json

{
        "Path": "github.com/x/y",
        "Main": true,
        "Dir": "/Users/kennethtan/Documents/Github/x/y,
        "GoMod": "/Users/kennethtan/Documents/Github/x/y/go.mod",
        "GoVersion": "1.18"
}
{
        "Path": "github.com/x/z",
        "Main": true,
        "Dir": "/Users/kennethtan/Documents/Github/x/z,
        "GoMod": "/Users/kennethtan/Documents/Github/x/z/go.mod",
        "GoVersion": "1.18"
}
```

Do note that the output are not delimited by commas which is an expected output as discussed (here)[https://github.com/golang/go/issues/27655]

Probably have to refactor the current code to 1) be able to properly parse the JSON output from stdout that do not have commas if there are more than one module and 2) iterate the list of modules to accurately return the list of imports required for rendering the dependency graph

**Comment by dustyhorizon** [comment]
Dependency graphs in go1.18 are not fully fixed as of nx-go 2.5.0.

go1.18 introduces workspaces which produces multiple modules when running `go list -m -json`

An example output of this is as follows:
```
$ go list -m -json

{
        "Path": "github.com/x/y",
        "Main": true,
        "Dir": "/Users/kennethtan/Documents/Github/x/y,
        "GoMod": "/Users/kennethtan/Documents/Github/x/y/go.mod",
        "GoVersion": "1.18"
}
{
        "Path": "github.com/x/z",
        "Main": true,
        "Dir": "/Users/kennethtan/Documents/Github/x/z,
        "GoMod": "/Users/kennethtan/Documents/Github/x/z/go.mod",
        "GoVersion": "1.18"
}
```

Do note that the output are not delimited by commas which is an expected output as discussed (here)[https://github.com/golang/go/issues/27655]

Probably have to refactor the current code to 1) be able to properly parse the JSON output from stdout that do not have commas if there are more than one module and 2) iterate the list of modules to accurately return the list of imports required for rendering the dependency graph

**Comment by Anthony-Jhoiro** [comment]
Any news on that @dustyhorizon, @phillipCouto ? Maybe I can help if needed

**Comment by phillipCouto** [comment]
> Any news on that @dustyhorizon, @phillipCouto ? Maybe I can help if needed

My apologizes, things got crazy for a bit and it completely slipped my radar. I will look into it today.

**Comment by phillipCouto** [comment]
Submitted a merge request that reworked the logic a bit to function in the traditional modules setup or a workspace. Also fixed a bug that missing imports to subfolders of projects and not properly establishing the relationship.

**Comment by beeman** [comment]
Thanks for your PRs again @phillipCouto! Highly appreciated. I'll look at merging them this week and cut a new release!

**Comment by beeman** [comment]
> What do you think about adding the possibility to create modules in a Go workspace (only for versions >1.18, of course)? For example, by having an option when generating an app or a lib to generate it with `go.mod` in the project directory and a `go.work` file in the root directory? It seems to be the most adapted way to handle monorepo with go, as having a single `go.mod` file is a bad practice in a project that contains multiple modules, as referenced in the [documentation](https://go.dev/doc/modules/managing-source#multiple-module-source). If you are interested, you would be happy to participate in writing the specs and the development!

Seems like a great addition to support Go Workspaces. 

Currently, I'm not using `nx-go` in any of the projects I'm actively working on, so I don't have the personal need, but feel free to spec it out! Definitely seems useful!

It would be great if @phillipCouto could chime in too as he has been contributing quite a bit to this project.

**Comment by phillipCouto** [comment]
> 



> > What do you think about adding the possibility to create modules in a Go workspace (only for versions >1.18, of course)? For example, by having an option when generating an app or a lib to generate it with `go.mod` in the project directory and a `go.work` file in the root directory? It seems to be the most adapted way to handle monorepo with go, as having a single `go.mod` file is a bad practice in a project that contains multiple modules, as referenced in the [documentation](https://go.dev/doc/modules/managing-source#multiple-module-source). If you are interested, you would be happy to participate in writing the specs and the development!
> 
> Seems like a great addition to support Go Workspaces.
> 
> Currently, I'm not using `nx-go` in any of the projects I'm actively working on, so I don't have the personal need, but feel free to spec it out! Definitely seems useful!
> 
> It would be great if @phillipCouto could chime in too as he has been contributing quite a bit to this project.

I agree @beeman it definitely would be a great enhancement to support it. Our projects are still pegged to go1.17 but once we upgrade to go1.18 we would definitely look at migrating to go workspaces.

@Anthony-Jhoiro if you got the time please spec it out in a separate issue and then we can go from there.

**Comment by phillipCouto** [comment]
> > @Anthony-Jhoiro, if you got the time, please spec it out in a separate issue, and then we can go from there.
> 
> I just created this issue #73. I will continue to update it in the next few days. @beeman @phillipCouto, I'd be happy to read your comments and thoughts 😉

I will take a look tonight :smile:.

**Comment by beeman** [comment]
This has been released as `@nx-go/nx-go@2.6.0` 🚀
